File: soc/bcm2710/display_dpi.c
/*
 * soc/bcm2710/display_dpi.c - BCM2710 Display Implementation (DPI/HDMI)
 *
 * Tutorial-OS: BCM2710 HAL Implementation
 *
 * Implements hal_display.h for BCM2710.
 * Existing framebuffer initialization code (fb_init_with_size) reorganized to fit the HAL structure.
 *
 * IMPORTANT: drawing functions (fb_clear, fb_draw_rect, etc.) are
 * completely portable and stay in drivers/framebuffer/framebuffer.c.
 * This file ONLY handles platform-specific initialization and buffer swap.
 */

#include "hal/hal_display.h"
#include "hal/hal_gpio.h"
#include "bcm2710_mailbox.h"

/* Include the portable framebuffer header for the struct definition */
/* This header stays in drivers/framebuffer/ */
#include "framebuffer.h"

/* =============================================================================
 * CONFIGURATION
 * =============================================================================
 */

#define FB_BITS_PER_PIXEL   32
#define FB_BUFFER_COUNT     2       /* Double buffering */
#define FB_DEFAULT_WIDTH    640
#define FB_DEFAULT_HEIGHT   480

/* =============================================================================
 * INTERNAL STATE
 * =============================================================================
 */

/* The actual framebuffer structure - matches existing framebuffer_t */
static framebuffer_t g_framebuffer;
static bool g_display_initialized = false;

/* =============================================================================
 * FRAMEBUFFER ALLOCATION VIA MAILBOX
 * =============================================================================
 * This is existing fb_init_with_size() mailbox code.
 */

static hal_error_t allocate_framebuffer(uint32_t width, uint32_t height)
{
    bcm_mailbox_buffer_t mbox = { .data = {0} };
    uint32_t virtual_height = height * FB_BUFFER_COUNT;

    /*
     * Build the framebuffer allocation request.
     */
    uint32_t i = 0;
    mbox.data[i++] = 0;                         /* Total size (filled at end) */
    mbox.data[i++] = BCM_MBOX_REQUEST;

    /* Set physical size */
    mbox.data[i++] = BCM_TAG_SET_PHYSICAL_SIZE;
    mbox.data[i++] = 8;
    mbox.data[i++] = 0;
    mbox.data[i++] = width;
    mbox.data[i++] = height;

    /* Set virtual size (height * buffer_count for double buffering) */
    mbox.data[i++] = BCM_TAG_SET_VIRTUAL_SIZE;
    mbox.data[i++] = 8;
    mbox.data[i++] = 0;
    mbox.data[i++] = width;
    mbox.data[i++] = virtual_height;

    /* Set virtual offset */
    mbox.data[i++] = BCM_TAG_SET_VIRTUAL_OFFSET;
    mbox.data[i++] = 8;
    mbox.data[i++] = 0;
    mbox.data[i++] = 0;
    mbox.data[i++] = 0;

    /* Set depth */
    mbox.data[i++] = BCM_TAG_SET_DEPTH;
    mbox.data[i++] = 4;
    mbox.data[i++] = 0;
    mbox.data[i++] = FB_BITS_PER_PIXEL;

    /* Set pixel order (BGR) */
    mbox.data[i++] = BCM_TAG_SET_PIXEL_ORDER;
    mbox.data[i++] = 4;
    mbox.data[i++] = 0;
    mbox.data[i++] = 0;                         /* 0 = BGR */

    /* Allocate buffer */
    mbox.data[i++] = BCM_TAG_ALLOCATE_BUFFER;
    mbox.data[i++] = 8;
    mbox.data[i++] = 0;
    uint32_t alloc_idx = i;
    mbox.data[i++] = 16;                        /* Alignment */
    mbox.data[i++] = 0;                         /* Size (output) */

    /* Get pitch */
    mbox.data[i++] = BCM_TAG_GET_PITCH;
    mbox.data[i++] = 4;
    mbox.data[i++] = 0;
    uint32_t pitch_idx = i;
    mbox.data[i++] = 0;                         /* Pitch (output) */

    /* End tag */
    mbox.data[i++] = BCM_TAG_END;

    /* Fill in total size */
    mbox.data[0] = i * 4;

    /* Send request */
    if (!bcm_mailbox_call(&mbox, BCM_MBOX_CH_PROP)) {
        return HAL_ERROR_DISPLAY_MAILBOX;
    }

    /* Extract results */
    uint32_t fb_addr = mbox.data[alloc_idx];
    uint32_t fb_size_bytes = mbox.data[alloc_idx + 1];
    uint32_t pitch = mbox.data[pitch_idx];

    if (fb_addr == 0 || fb_size_bytes == 0) {
        return HAL_ERROR_DISPLAY_NO_FB;
    }

    /* Convert bus address to ARM address */
    fb_addr = BCM_BUS_TO_ARM(fb_addr);

    /* Fill in framebuffer structure */
    g_framebuffer.width = width;
    g_framebuffer.height = height;
    g_framebuffer.pitch = pitch;
    g_framebuffer.virtual_height = virtual_height;
    g_framebuffer.buffer_size = pitch * height;

    g_framebuffer.buffers[0] = (uint32_t *)(uintptr_t)fb_addr;
    if (FB_BUFFER_COUNT > 1) {
        g_framebuffer.buffers[1] = (uint32_t *)(uintptr_t)(fb_addr + g_framebuffer.buffer_size);
    }

    g_framebuffer.front_buffer = 0;
    g_framebuffer.back_buffer = (FB_BUFFER_COUNT > 1) ? 1 : 0;
    g_framebuffer.addr = g_framebuffer.buffers[g_framebuffer.back_buffer];

    /* Initialize clipping to full screen */
    g_framebuffer.clip_depth = 0;
    g_framebuffer.clip_stack[0].x = 0;
    g_framebuffer.clip_stack[0].y = 0;
    g_framebuffer.clip_stack[0].w = width;
    g_framebuffer.clip_stack[0].h = height;

    /* Initialize dirty tracking */
    g_framebuffer.dirty_count = 0;
    g_framebuffer.full_dirty = true;
    g_framebuffer.frame_count = 0;
    g_framebuffer.vsync_enabled = true;
    g_framebuffer.initialized = true;

    return HAL_SUCCESS;
}

/* =============================================================================
 * HAL DISPLAY INITIALIZATION
 * =============================================================================
 */

hal_error_t hal_display_init(framebuffer_t **fb_out)
{
    return hal_display_init_with_size(FB_DEFAULT_WIDTH, FB_DEFAULT_HEIGHT, fb_out);
}

hal_error_t hal_display_init_with_size(uint32_t width, uint32_t height,
                                        framebuffer_t **fb_out)
{
    if (fb_out == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    if (g_display_initialized) {
        return HAL_ERROR_ALREADY_INIT;
    }

    /* Configure GPIO for DPI (if using DPI output like GPi Case) */
    hal_error_t err = hal_gpio_configure_dpi();
    if (HAL_FAILED(err)) {
        /* DPI configuration failed - might be HDMI only board */
        /* Continue anyway, let mailbox handle it */
    }

    /* Allocate framebuffer via mailbox */
    err = allocate_framebuffer(width, height);
    if (HAL_FAILED(err)) {
        return err;
    }

    g_display_initialized = true;
    *fb_out = &g_framebuffer;

    return HAL_SUCCESS;
}

hal_error_t hal_display_init_with_config(const hal_display_config_t *config,
                                          framebuffer_t **fb_out)
{
    if (config == NULL || fb_out == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    /* For now, just use width/height from config */
    return hal_display_init_with_size(config->width, config->height, fb_out);
}

hal_error_t hal_display_shutdown(void)
{
    if (!g_display_initialized) {
        return HAL_ERROR_NOT_INIT;
    }

    /* Can't really deallocate the framebuffer on BCM */
    g_display_initialized = false;
    g_framebuffer.initialized = false;

    return HAL_SUCCESS;
}

bool hal_display_is_initialized(void)
{
    return g_display_initialized;
}

/* =============================================================================
 * DISPLAY INFORMATION
 * =============================================================================
 */

hal_error_t hal_display_get_info(hal_display_info_t *info)
{
    if (info == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    if (!g_display_initialized) {
        return HAL_ERROR_NOT_INIT;
    }

    info->width = g_framebuffer.width;
    info->height = g_framebuffer.height;
    info->pitch = g_framebuffer.pitch;
    info->bits_per_pixel = FB_BITS_PER_PIXEL;
    info->type = HAL_DISPLAY_TYPE_DPI;
    info->format = HAL_PIXEL_FORMAT_ARGB8888;
    info->framebuffer_base = g_framebuffer.buffers[0];
    info->framebuffer_size = g_framebuffer.buffer_size * FB_BUFFER_COUNT;
    info->buffer_count = FB_BUFFER_COUNT;

    return HAL_SUCCESS;
}

uint32_t hal_display_get_width(void)
{
    return g_display_initialized ? g_framebuffer.width : 0;
}

uint32_t hal_display_get_height(void)
{
    return g_display_initialized ? g_framebuffer.height : 0;
}

uint32_t hal_display_get_pitch(void)
{
    return g_display_initialized ? g_framebuffer.pitch : 0;
}

/* =============================================================================
 * DISPLAY CONTROL
 * =============================================================================
 * These match existing fb_present() and fb_set_vsync() functions.
 */

hal_error_t hal_display_present(framebuffer_t *fb)
{
    if (fb == NULL || !fb->initialized) {
        return HAL_ERROR_NOT_INIT;
    }

    /* Wait for vsync if enabled */
    if (fb->vsync_enabled) {
        bcm_mailbox_wait_vsync();
    }

    /* Swap buffers */
    if (FB_BUFFER_COUNT > 1) {
        uint32_t temp = fb->front_buffer;
        fb->front_buffer = fb->back_buffer;
        fb->back_buffer = temp;
        fb->addr = fb->buffers[fb->back_buffer];
    }

    /* Set virtual offset to display front buffer */
    uint32_t y_offset = fb->front_buffer * fb->height;
    bcm_mailbox_set_virtual_offset(0, y_offset);

    fb->frame_count++;
    fb->full_dirty = false;
    fb->dirty_count = 0;

    return HAL_SUCCESS;
}

hal_error_t hal_display_present_immediate(framebuffer_t *fb)
{
    if (fb == NULL || !fb->initialized) {
        return HAL_ERROR_NOT_INIT;
    }

    /* Same as present but skip vsync */
    if (FB_BUFFER_COUNT > 1) {
        uint32_t temp = fb->front_buffer;
        fb->front_buffer = fb->back_buffer;
        fb->back_buffer = temp;
        fb->addr = fb->buffers[fb->back_buffer];
    }

    uint32_t y_offset = fb->front_buffer * fb->height;
    bcm_mailbox_set_virtual_offset(0, y_offset);

    fb->frame_count++;
    fb->full_dirty = false;
    fb->dirty_count = 0;

    return HAL_SUCCESS;
}

hal_error_t hal_display_set_vsync(framebuffer_t *fb, bool enabled)
{
    if (fb == NULL || !fb->initialized) {
        return HAL_ERROR_NOT_INIT;
    }

    fb->vsync_enabled = enabled;
    return HAL_SUCCESS;
}

hal_error_t hal_display_wait_vsync(void)
{
    if (bcm_mailbox_wait_vsync()) {
        return HAL_SUCCESS;
    }
    return HAL_ERROR_HARDWARE;
}

/* =============================================================================
 * DEFAULT CONFIGURATION
 * =============================================================================
 */

hal_error_t hal_display_get_default_config(hal_display_config_t *config)
{
    if (config == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    /* Default for GPi Case 2W */
    config->width = FB_DEFAULT_WIDTH;
    config->height = FB_DEFAULT_HEIGHT;
    config->type = HAL_DISPLAY_TYPE_DPI;
    config->format = HAL_PIXEL_FORMAT_ARGB8888;
    config->buffer_count = FB_BUFFER_COUNT;
    config->vsync_enabled = true;

    return HAL_SUCCESS;
}

File: soc/bcm2710/src/display_dpi.rs
//! BCM2710 display implementation (DPI / HDMI via VideoCore mailbox).
//!
//! Only initialization and buffer presentation are platform-specific.
//! All drawing primitives live in the portable `drivers::framebuffer` crate.

use hal::display::{Display, DisplayConfig, DisplayType, FramebufferInfo, PixelFormat};
use hal::gpio::Peripheral;
use hal::types::{HalError, HalResult};
use crate::{mailbox, regs};

const BITS_PER_PIXEL: u32 = 32;
const BUFFER_COUNT: u32   = 2;
const DEFAULT_WIDTH: u32  = 640;
const DEFAULT_HEIGHT: u32 = 480;

/// BCM2710 display controller.
pub struct Bcm2710Display {
    info: Option<FramebufferInfo>,
    /// Index of the buffer currently shown on screen.
    front: u32,
    /// Index of the buffer the CPU is drawing into.
    back: u32,
    height: u32,
    vsync: bool,
}

impl Bcm2710Display {
    pub const fn new() -> Self {
        Self {
            info: None,
            front: 0,
            back: 1,
            height: 0,
            vsync: true,
        }
    }

    /// Allocate a framebuffer through the VideoCore mailbox.
    fn allocate(&mut self, width: u32, height: u32) -> HalResult<FramebufferInfo> {
        let virtual_height = height * BUFFER_COUNT;

        let mut mbox = mailbox::MailboxBuffer::zeroed();
        let mut i = 0usize;

        // Build the multi-tag request.
        mbox.data[i] = 0;                      i += 1; // total size (filled at end)
        mbox.data[i] = regs::MBOX_REQUEST;      i += 1;

        // Set physical size
        mbox.data[i] = regs::tag::SET_PHYSICAL_SIZE; i += 1;
        mbox.data[i] = 8;  i += 1;
        mbox.data[i] = 0;  i += 1;
        mbox.data[i] = width;  i += 1;
        mbox.data[i] = height; i += 1;

        // Set virtual size
        mbox.data[i] = regs::tag::SET_VIRTUAL_SIZE; i += 1;
        mbox.data[i] = 8;  i += 1;
        mbox.data[i] = 0;  i += 1;
        mbox.data[i] = width;          i += 1;
        mbox.data[i] = virtual_height; i += 1;

        // Set virtual offset
        mbox.data[i] = regs::tag::SET_VIRTUAL_OFFSET; i += 1;
        mbox.data[i] = 8;  i += 1;
        mbox.data[i] = 0;  i += 1;
        mbox.data[i] = 0;  i += 1;
        mbox.data[i] = 0;  i += 1;

        // Set depth
        mbox.data[i] = regs::tag::SET_DEPTH; i += 1;
        mbox.data[i] = 4;  i += 1;
        mbox.data[i] = 0;  i += 1;
        mbox.data[i] = BITS_PER_PIXEL; i += 1;

        // Set pixel order (BGR)
        mbox.data[i] = regs::tag::SET_PIXEL_ORDER; i += 1;
        mbox.data[i] = 4;  i += 1;
        mbox.data[i] = 0;  i += 1;
        mbox.data[i] = 0;  i += 1; // 0 = BGR

        // Allocate buffer
        mbox.data[i] = regs::tag::ALLOCATE_BUFFER; i += 1;
        mbox.data[i] = 8;  i += 1;
        mbox.data[i] = 0;  i += 1;
        let alloc_idx = i;
        mbox.data[i] = 16; i += 1; // alignment
        mbox.data[i] = 0;  i += 1; // size (output)

        // Get pitch
        mbox.data[i] = regs::tag::GET_PITCH; i += 1;
        mbox.data[i] = 4;  i += 1;
        mbox.data[i] = 0;  i += 1;
        let pitch_idx = i;
        mbox.data[i] = 0;  i += 1; // pitch (output)

        // End tag
        mbox.data[i] = regs::tag::END; i += 1;

        // Fill total size
        mbox.data[0] = (i as u32) * 4;

        if !mailbox::call(&mut mbox, regs::MBOX_CH_PROP) {
            return Err(HalError::DisplayMailbox);
        }

        let fb_addr = mbox.data[alloc_idx];
        let fb_size = mbox.data[alloc_idx + 1];
        let pitch   = mbox.data[pitch_idx];

        if fb_addr == 0 || fb_size == 0 {
            return Err(HalError::DisplayNoFramebuffer);
        }

        // Convert bus address → ARM physical address
        let fb_addr = regs::bus_to_arm(fb_addr);

        self.height = height;
        self.front = 0;
        self.back = if BUFFER_COUNT > 1 { 1 } else { 0 };

        let info = FramebufferInfo {
            width,
            height,
            pitch,
            bits_per_pixel: BITS_PER_PIXEL,
            display_type: DisplayType::Dpi,
            format: PixelFormat::Argb8888,
            base: fb_addr as *mut u32,
            size: fb_size,
            buffer_count: BUFFER_COUNT,
        };

        self.info = Some(info);
        Ok(info)
    }
}

impl Display for Bcm2710Display {
    fn init(&mut self, config: Option<&DisplayConfig>) -> HalResult<FramebufferInfo> {
        if self.info.is_some() {
            return Err(HalError::AlreadyInitialized);
        }

        let (w, h) = match config {
            Some(c) => (c.width, c.height),
            None    => (DEFAULT_WIDTH, DEFAULT_HEIGHT),
        };

        // Configure GPIO for DPI (best-effort — HDMI-only boards skip this)
        let gpio = crate::gpio::Bcm2710Gpio;
        let _ = hal::gpio::Gpio::configure_peripheral(&gpio, Peripheral::Dpi);

        self.allocate(w, h)
    }

    fn shutdown(&mut self) -> HalResult<()> {
        if self.info.is_none() {
            return Err(HalError::NotInitialized);
        }
        // Can't deallocate the framebuffer on BCM — just mark uninitialised.
        self.info = None;
        Ok(())
    }

    fn is_initialized(&self) -> bool { self.info.is_some() }

    fn width(&self)  -> u32 { self.info.map(|i| i.width).unwrap_or(0) }
    fn height(&self) -> u32 { self.info.map(|i| i.height).unwrap_or(0) }
    fn pitch(&self)  -> u32 { self.info.map(|i| i.pitch).unwrap_or(0) }

    fn present(&mut self) -> HalResult<u32> {
        if self.info.is_none() { return Err(HalError::NotInitialized); }

        if self.vsync {
            mailbox::wait_vsync();
        }

        // Swap front ↔ back
        if BUFFER_COUNT > 1 {
            core::mem::swap(&mut self.front, &mut self.back);
        }

        // Tell the GPU to display the front buffer
        let y_offset = self.front * self.height;
        mailbox::set_virtual_offset(0, y_offset);

        Ok(self.back)
    }

    fn present_immediate(&mut self) -> HalResult<u32> {
        let saved = self.vsync;
        self.vsync = false;
        let result = self.present();
        self.vsync = saved;
        result
    }

    fn set_vsync(&mut self, enabled: bool) -> HalResult<()> {
        self.vsync = enabled;
        Ok(())
    }

    fn wait_vsync(&self) -> HalResult<()> {
        if mailbox::wait_vsync() { Ok(()) } else { Err(HalError::HardwareFault) }
    }

    fn default_config(&self) -> DisplayConfig {
        DisplayConfig {
            width: DEFAULT_WIDTH,
            height: DEFAULT_HEIGHT,
            display_type: DisplayType::Dpi,
            format: PixelFormat::Argb8888,
            buffer_count: BUFFER_COUNT,
            vsync: true,
        }
    }
}
